### PR TITLE
Bugfixes and speedup

### PR DESCRIPTION
**Pull request by mv3shape**
I've been testing Obfuscar on a large project with 70+ assemblies, lots of XAML with styling and lots of generics, and in order to make the obfuscation pass without crashing and get a working application afterwards I've fixed a number of bugs. Also, the process now runs in about half the time, from ~2mins to ~1min on my app. All tests pass, including a few new ones I added.

### CODE REVIEW

**Comment by casz** [comment]
FYI, I also signed the CLA as I was so kind to cleanup some of @mv3shape commit history 😅 
When we forked the repo on GitHub and moved our changes to the GitHub repo 👍

**Comment by casz** [comment]
👋 @SorenMaagaard please remember to sign the CLA

File: Obfuscar/AssemblyInfo.cs [hunk 1]
```diff
@@ -1071,6 +1096,25 @@ public bool ShouldSkipStringHiding(MethodKey method, InheritMap map, bool projec
                 return true;
             }
 
+            if (typeInXaml && prop.Property.IsPublic())
+            {
+                message = "filtered by BAML";
+                return true;
+            }
+
+            if (prop.Property.IsPublic() && prop.DeclaringType.ImplementsInterface("System.ComponentModel.INotifyPropertyChanged"))
```
Comment (lextm) [inline Obfuscar/AssemblyInfo.cs]: Not all public properties of a type derived from INotifyPropertyChanged are used for MVVM pattern (though I agree most of them should), so I still prefer explicit skip rules to such automatic processing.

**Comment by lextm** [comment]
I think it should be split into several pull requests for me to review. Some of the changes (logging related) are trivial while others can be further discussed.

I cannot take it in the current form.

**Comment by mv3shape** [comment]
I have fixed around 20 different issues primarily related to XAML and using complex generics across multiple assemblies, but also simpler things like being able to override Equals and GetHashCode. There is more or less one fix in each commit as you can see on the commit descriptions, but changing it to 20 independent pull requests would be a lot of work as some of the fixes also build on each other. 
As you point out some of the things with XAML bindings could be handled with skip rules or attributes, but this is easy to forget and hard to maintain, especially because it often requires manual testing of the obfuscate app to find such mistakes.
We are a large organization with more than 400 developers working on a large number of projects which also share a lot of code. We would like a streamlined build pipeline where the output is obfuscated but without everyone having to be obfuscation experts, Therefore we need to be able to make a fairly safe configuration where obfuscation doesn't introduce unexpected errors, perhaps at the cost of a few properties not being obfuscated even though they could have been.

File: Obfuscar/AssemblyInfo.cs (hunk 1 above)
Comment (mv3shape) [reply Obfuscar/AssemblyInfo.cs]: What do you think about making it an option like I did with AbortOnInconsistentState?

**Comment by lextm** [comment]
Thanks. Then I might have to cherry pick the commits, instead of accepting them in this pull request. I will close this one once I finish reviewing each commits.

**Comment by mv3shape** [comment]
Of course, if you think I've made a mistake let me know so that I can correct it and update the request..

**Comment by GF-Huang** [comment]
What's the progress?

**Comment by lextm** [comment]
@GF-Huang you can try 3sharp's fork out yourself https://github.com/3shape/obfuscar

I don't plan to accept this pull request yet, because at least it doesn't work for a few key projects of my own.

**Comment by GF-Huang** [comment]
> @GF-Huang you can try 3sharp's fork out yourself https://github.com/3shape/obfuscar
> 
> I don't plan to accept this pull request yet, because at least it doesn't work for a few key projects of my own.

How about https://github.com/obfuscar/obfuscar/issues/58? Any plans?

**Comment by mv3shape** [comment]
@lextm which problems do you experience with your projects?

**Comment by lextm** [comment]
@mv3shape They are relatively large closed source projects, so I won't be able to share much information.

**Pull request closed without merging**